MergeScenes node

johnhaddon opened the pull request:
This implements #3583. It's one of our more complicated nodes to date, so would be good to get lots of eyes on it, both from an implementation standpoint and a design one.

While implementing this, and in discussion with Tom, I made a couple of calls on details that weren't completely specified in the discussion on #3583 :

- I interpreted the toggles as not having any effect where a location exists in only one input scene. So new bits of hierarchy from additional locations are always merged in with all their properties intact, regardless of toggles. So toggles are only relevant where two inputs have the same location. I'm pretty sure this is what everyone was assuming.
- I didn't add a toggle for disabling the merging of additional items of hierarchy. This would have pretty dire performance consequences for set computations, because we'd have to traverse the hierarchy to figure out what we needed to prune from the sets before merging them.
- I didn't add a toggle for disabling the merging of sets. Similar to the above, it would entail traversing the hierarchy to figure out which the _new_ bits were so we could keep those bits of the sets and not the others.
- I took "merge objects" to mean "pick the rightmost one", not "hope the primitive types are compatible and keep the elements of both".
- I defaulted all the toggles to off, so the default state of the node is to behave like the simpler MergeHierarchy idea. I did this because I think it still covers a lot of use cases in this state, and turning any of the toggles on has a performance impact (especially for objects/transforms, because they kick us into a worst-case path for bounds computations). I do worry that someone might not immediately understand that they still get useful hierarchy merging without turning anything on, and get in the habit of just flipping everything on. Is it worth having an extra top-level mode thing that switches between "Merge Hierarchy" and "Merge Properties", so you only get access to the toggles by switching mode?

Hopefully that all makes sense.

danieldresser-ie commented:
Hmm, started trying to take a look at this, but didn't get far.

It currently fails to compile at IE because of this bug in our old gcc 4.8.3:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=59483

themissingcow commented on include/GafferScene/MergeScenes.h:
```diff
@@ -0,0 +1,141 @@
+		const Gaffer::BoolPlug *objectsPlug() const;
+
+		Gaffer::BoolPlug *globalsPlug();
+		const Gaffer::BoolPlug *globalsPlug() const;
```
I think personally I'd prefer `merge*Plug` as it saves a few double takes when following the general logic in affects/hash/compute. It think it's also a bit clearer when working with the node in python if the node itself isn't called something "mergey".

johnhaddon replied:
Makes sense, but given that we've decided to introduce a mode, I've renamed to `*ModePlug` in b8c001a67e88d3fdd957e878ef00c984590e4a0a. This fits with our mode plugs on other nodes.

themissingcow commented on python/GafferSceneTest/MergeScenesTest.py:
```diff
@@ -0,0 +1,364 @@
+			 # `SceneNode::hashTransform()` isn't called for the root.
+			 checks = self.allSceneChecks - { "transform" }
+		)
+
```
Be good to add a no-first-input test.

johnhaddon replied:
6e45cfeb2933d3e5e564dd84a7c964b51d1c883c

themissingcow commented on python/GafferSceneTest/MergeScenesTest.py:
```diff
@@ -0,0 +1,364 @@
+
+		merge = GafferScene.MergeScenes()
+		merge["in"][0].setInput( sphere["out"] )
+		merge["in"][1].setInput( group["out"] )
```
`merge["objects"].setValue( True )` ?

johnhaddon replied:
https://github.com/GafferHQ/gaffer/pull/3642/commits/de00858529f8e21c2b6d380580c21716bc9d84df

themissingcow commented on python/GafferSceneUI/MergeScenesUI.py:
```diff
@@ -0,0 +1,126 @@
+	Optionally, where multiple inputs contain the same location,
+	the properties at that location may also be merged. This
+	merging is enabled on a per-property basis.
+
```
Worth adding a note that sets are always merged?

johnhaddon replied:
https://github.com/GafferHQ/gaffer/pull/3642/commits/9fbdc6847c958f103ea80760eb1679dd4efaec55

johnhaddon commented:
Quick note about the outcome of a quick demo I did with a couple of lighters :

- The toggles were a bit baffling. It wasn't obvious that something would still happen even when they were off (new locations would be merged) . It wasn't obvious what "On" meant, or why transforms and objects followed different rules (last wins) vs attributes and globals (merge them).
- They were enthusiastic about the functionality once they'd figured it out.

I plan to resolve this as follows :

- Replace the transforms and objects toggles with dropdown menus with values of `First, Last`, corresponding in behaviour to the current `true, false` values.
- Replace the attributes and globals toggles with a dropdown menu with values of `First, Last, Merged`. First and merged correspond to the current `false, true` values, and `Last` just passes through the last input which is useful if you want to obliterate old lookdev with new.

Hopefully this makes it clear that things still happen when the plugs are at their default values, and that the object and transform modes are simple switches and not a fancy merge. It also opens the possibility of adding other merging modes in future without any backwards compatibility headaches (we just add new modes to the existing menus).

themissingcow commented:
> * Replace the transforms and objects toggles with dropdown menus with values of `First, Last`, corresponding in behaviour to the current `true, false` values.
> * Replace the attributes and globals toggles with a dropdown menu with values of `First, Last, Merged`. First and merged correspond to the current `false, true` values, and `Last` just passes through the last input which is useful if you want to obliterate old lookdev with new.

I like this, it explains how it works allowing you to pick the right option, rather than relying on you already knowing how it works.

johnhaddon commented:
> It currently fails to compile at IE because of this bug in our old gcc 4.8.3:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=59483

![image](https://user-images.githubusercontent.com/1133871/74960092-9afb0a80-5403-11ea-90b8-626ffe40da82.png)

Ugh. Remind me, why do we have a floor of GCC 4.8 and C++11 when VFXPlatform has long since left them behind?

I can probably rustle up a GCC 4.8 development environment to take a look at this tomorrow, but in the meantime perhaps you can just put `#define protected public` at the top of MergeScenes.cpp? Unless there's another minimally intrusive workaround, I suspect we'll end up just doing that but protected by an `#if GCC_VERSION < 4.9` anyway.

andrewkaufman commented:
> why do we have a floor of GCC 4.8 and C++11 when VFXPlatform has long since left them behind?

Because Maya 2018 is still used in production, and Nuke still has not released a gcc6 version.

danieldresser-ie commented:
Talking with John about the UI, I definitely agree that the new proposal is a lot better than the checkboxes.

The main thing that bugs me is that First/Last implies a symmetry that doesn't fully exist:
I would expect that setting everything to First and plugging node A into the first input and node B into the second would yield the same result as setting everything to Last and plugging node B in the first input and node B into the second.

I proposed the possible terminology "Keep/Merge/Replace" instead of "First/Merge/Last".  I like it a fair bit better, and it sounds like John thought it might be reasonable.  It reduces the expectation for symmetry, and in my opinion, also makes it easier to remember that "Keep" is doing less work than "Replace" ( whereas First/Last sound symmetrical, so it's surprising that one is much more expensive than the other ).

danieldresser-ie commented:
Couple of good comments from Andrew:
* he pointed out that I hadn't actually described why the processing of First/Last isn't symmetric.  I told him that Last takes the last non-null object, but first takes the first object that exists at the location
* he agreed that he found this confusing, but then threw another wrench in the works:  what if you actually want to remove objects that aren't found in the scene you're replacing with?  Do we actually need 4 modes: "First/Last/Merge/Replace"?  Where both First and Last are fast paths that take the first or last regardless of the value, and replace is the more expensive mode that checks the value of the object, and only replaces if there is an object there?

danieldresser-ie commented:
And sorry, didn't really get too far with the code review, I'll set aside more time tomorrow for digging into this.

johnhaddon commented:
> what if you actually want to remove objects that aren't found in the scene you're replacing with? Do we actually need 4 modes: "First/Last/Merge/Replace"? Where both First and Last are fast paths that take the first or last regardless of the value, and replace is the more expensive mode that checks the value of the object, and only replaces if there is an object there?

Is this hypothetical? I don't want to add options that aren't useful and which make the common operation of the node harder to understand. I was convinced by your Keep/Replace/Merge suggestion, and without a non-hypothetical use case I'm inclined to say that First/Last/Merge/Replace just reintroduces some of the original confusion and adds a bit more.

The closest I can come to a use case is wanting to be able to apply the output of a hypothetical SceneDiff node such that the operation of DeleteObjects can be recorded and played back. But this `Last` mode doesn't achieve that, because we have no way of knowing if we've got a NullObject due to deletion or a lack of differences. I have some thoughts about this, but they're future work, and won't benefit from a `Last` mode.

johnhaddon commented on src/GafferScene/MergeScenes.cpp:
```diff
@@ -0,0 +1,944 @@
+
+Gaffer::BoolPlug *MergeScenes::adjustBoundsPlug()
+{
+	return getChild<BoolPlug>( g_firstPlugIndex + 4);
```
067b45002695dab3d1253f4fa6f2f754b4c9f99d

johnhaddon commented on include/GafferScene/MergeScenes.h:
```diff
@@ -0,0 +1,141 @@
+		void visitActiveInputs( const ScenePath &path, const Gaffer::Context *context, Visitor &&visitor, VisitOrder order = VisitOrder::Forwards ) const;
+
+		template<typename Visitor>
+		void visitActiveInputs( Visitor &&visitor ) const;
```
Ironically I had originally called it `visitConnectedInputs()` and then renamed it when I added the "always visits at least one input, even if not connected" behaviour. But I think you're right, it needs to be clarified : 204b16fba2daea72c0ca49b87748679f741636b4.

johnhaddon commented on src/GafferScene/MergeScenes.cpp:
```diff
@@ -0,0 +1,944 @@
+		{
+			visitor( type, i, scene.get() );
+		}
+		type = InputType::Other;
```
Good catch! 6e45cfeb2933d3e5e564dd84a7c964b51d1c883c.

johnhaddon commented on src/GafferScene/MergeScenes.cpp:
```diff
@@ -0,0 +1,944 @@
+	const auto &path = context->get<ScenePlug::ScenePath>( ScenePlug::scenePathContextName );
+	if( path.empty() )
+	{
+		// Root
```
Somewhat improved by 204b16fba2daea72c0ca49b87748679f741636b4.

johnhaddon commented on src/GafferScene/MergeScenes.cpp:
```diff
@@ -0,0 +1,944 @@
+					SceneProcessor::hashObject( path, context, parent, h );
+				case InputType::Other :
+					// Merge input hash.
+					scene->objectPlug()->hash( h );
```
Correct. There's no way of knowing what type the object will be without actually computing the object, so this is the best we can do.

johnhaddon commented on src/GafferScene/MergeScenes.cpp:
```diff
@@ -0,0 +1,944 @@
+
+IECore::ConstInternedStringVectorDataPtr MergeScenes::computeChildNames( const ScenePath &path, const Gaffer::Context *context, const ScenePlug *parent ) const
+{
+	return mapping()->outputChildNames();
```
The split in `hashChildNames()` is to satisfy the unit tests that assert that we have a perfect pass-through of the hash. In `computeChildNames()` we could have the same special case, but it wouldn't bring any real benefit - both `mapping()->outputChildNames()` and `inPlugs( soleInputWithChildren() )->childNamesPlug()->getValue()` reference the same underlying data anyway, thanks to copy on write.

johnhaddon commented:
@themissingcow, I've added the Mode enum in b8c001a, and addressed your other comments in separate commits. Just got to take a look at this GCC 4.8 nonsense now.

@andrewkaufman, @danieldresser-ie, I've made a series of fixup commits to make it easier for Tom to verify, but if you're just starting a review I think it'll be easier to just look at the overall changes.

johnhaddon commented:
> It currently fails to compile at IE because of this bug in our old gcc 4.8.3:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=59483

Added workaround in 9c7f6e9.

danieldresser commented:
Sorry it's taken me a while to comment on this, I've been trying to wrap my head around it.

It does feel like the data flow is a bit more confusing than it should be.  The main things that feel to me to be confusing are:

* visit*Inputs is used both to compute mappingPlug, and also to quickly retreive the value of mappingPlug
* there are a bunch of places where we need to check if path is empty, and call the other variant of visit*Inputs, despite the fact that visitActiveInputs has a branch for path being empty inside it

If we look at the most common use case for mappingPlug, we just need to check the activeInputs for the current location.  This could be acheived much more simply if we just stored nothing but m_activeInputs based on the current path in the context.  If we moved the code for computing this simle version of mappingPlug out of visitActiveInputs, directly into computeMapping, it seems like we could be left with a very simple visitActiveInputs which is safe to call whether or not path is empty.  Maybe this is oversimplifying things, but the fact that currently the only use of childActiveInputs is when we create a special context to get the parent, and then ask what the active inputs of the child with our current name are, makes me really think that there should be a simpler path where we just query the current location.

What this approach doesn't address is the places where we use mappingPlug() outside visit*Inputs, namely the bounds and childNames computations.  It's not clear to me how much we're saving by computing the childNames inside the mappingPlug instead of computing them when childNames is queried.  It's definitely possible that we would need a second, more complex cache - but I think it could be computed as a more complex second layer, using the simplifed visitActiveInputs that queries a simple mappingPlug()?  ( Instead of the compute process for the more complicated data getting tangled up in computing the cache that is needed to make visitActiveInputs work? )

Anyway, this is complex enough that I'm quite likely missing something, but I thought I should say it in case this gives you some sort of idea on the structure.  Oh, and I can't resist commenting on your question about bit twiddling, which I love, but unfortunately, I don't think there's anything really succinct and convenient for finding index of the lowest bit set, unless you use something like builtin_ctz ( which corresponds to an actual processor instruction ), which might be a headache across different compilers.

johnhaddon commented:
Hey Daniel,

Thanks for the feedback...

> If we look at the most common use case for mappingPlug, we just need to check the activeInputs for the current location. This could be acheived much more simply if we just stored nothing but m_activeInputs based on the current path in the context.

I may not be following correctly, but let me pretend I am and we'll see if it gets us anywhere. I'm hoping that the problem stems from me not having commented MappingData thoroughly enough.

The reason for MappingData is performance. Think about how you'd compute `m_activeInputs` _directly_ for a location, without help from a parent-level mapping. You'd need to know if the current location existed for each input, and the only way to get that would be via `SceneAlgo::exists()` or an equivalent. `SceneAlgo::exists()` is slow, because it has to start at the root _every time_, and at each level from the root down it has to do a linear search in `childNames`. MappingData fixes both of those problems. First, when computing MappingData at a child location, we only ever consider the active inputs from the parent mapping. For merges of mostly disjoint scenes this means we quickly filter down into the `soleInputWithChildren()` case for the majority of locations, with very little work to do at each of the descendants. Second, `m_children` means we never do that linear search. We can't avoid linear search by considering children one-by-one, we need a data structure held one level up on the parent.

Now, I have been thinking that `SceneAlgo::exists()` should really be backed by a parent-level structure anyway, but even with that, MappingData would still have an edge. With only `exists()` we would still need to check all inputs at all locations, whereas with MappingData we very quickly drill down to the we-know-only-a-single-input-is-valid-for-all-descendants case.

> It's definitely possible that we would need a second, more complex cache - but I think it could be computed as a more complex second layer, using the simplifed visitActiveInputs that queries a simple mappingPlug()? 

Hopefully I've convinced you that this is back to front - to get the simple `m_activeInputs` with any kind of performance, you need to derive it from a _first_, more complex cache. That's what `MappingData` is. Since we need `m_children` for fast lookups, and we need fast lookups for merging `childNames`, it made sense to me to compute `childNames` in `MappingData` too, rather than make the dataflow any more complex.

> there are a bunch of places where we need to check if path is empty, and call the other variant of visit*Inputs, despite the fact that visitActiveInputs has a branch for path being empty inside it

To be precise, there are three. Two of them are the hash/compute pair for the mapping, so they count as one really. The mapping is computed recursively (each child depends on the parent), so there must be a special case where the recursion bottoms out. The other one is in `visitActiveInputs()`, and is necessary because there is obviously no parent to retrieve the mapping from when we're at the root. We could eradicate that check with a nice tidy rule that everyone should call `visitConnected()` for root locations and `visitActive()` for other locations. But that's a mess in reality, because it moves duplicates of the check into all the `visitActive()` call sites.

Hopefully I'm making sense. If not, maybe it's worth moving our weekly chat forwards this week?
Cheers...
John

danieldresser-ie commented:
Oooh, right.  I should have thought of that.  Forgot that checking if a location exists is totally accursed.

This still feels a bit horrifyingly complicated for what it's doing, but you're probably right that there isn't any simpler option that doesn't sacrifice performance.

On the positive side, if this does actually offer a way to check if a  location exists with acceptable perfcormance, that would be great news.  It drives me crazy that the IE denoiser currently can't check if locations exist on adjacent frames - this still regularly kills production renders at the worst possible time.

johnhaddon commented on src/GafferScene/MergeScenes.cpp:
```diff
@@ -0,0 +1,1084 @@
+			}
+
+			// Not the first input with children. Move any existing sole input into
+			// m_children if necessary.
```
It's meaningless : e9b4625.

johnhaddon commented:
> This still feels a bit horrifyingly complicated for what it's doing, but you're probably right that there isn't any simpler option that doesn't sacrifice performance.

To paraphrase Einstein, it's as horrifyingly complicated as necessary, but no more horrifyingly complicated. Just be glad we don't have the horror of giving this thing multiple merge points using a filter.

> On the positive side, if this does actually offer a way to check if a location exists with acceptable perfcormance, that would be great news.

I was thinking it'd look something like this :

- An `InternedStringVectorData *ScenePlug::sortedChildNamesPlug()` (possibly private).
- A `BoolPlug *ScenePlug::existsPlug()`.
- Both computed automatically by the SceneNode base class, with `computeExists()` using the same recursive strategy that MappingData uses - `exists( parent ) && childName in sortedChildNames( parent )`.

Actually, I guess if we had that, it should be possible to derive a simpler `MergeScenes::activeInputsPlug()` from just `exists( child ) for child in activeInputs( parent )` and then compute child names separately. Not sure if there's some lurking gotcha that might bite if I tried to do that though.

johnhaddon commented:
> Actually, I guess if we had that [ScenePlug::existsPlug()], it should be possible to derive a simpler MergeScenes::activeInputsPlug() from just `exists( child ) for child in activeInputs( parent )` and then compute child names separately. Not sure if there's some lurking gotcha that might bite if I tried to do that though.

I've been hacking away today to prototype this approach and it does seem to work reasonably well. If you're interested in taking a look here's the code :

https://github.com/johnhaddon/gaffer/tree/mergeScenesTesting

Observations so far :

- It definitely results in simpler MergeScenes code.
- There is some overhead against the version in the PR (about 6% on a test that merge two big Instancer hierarchies).
- The overhead is minimal compared to the naive approach of using the old `SceneAlgo::exists()` at every location. That approach has more than 100% overhead.
- The optimised existence queries knock 52% off a CopyPrimitiveVariables benchmark (our heaviest pre-existing user of `exists()`).
- The new `ScenePlug::existsPlug()` and `ScenePlug::sortedChildNamesPlug()` feel a little bit odd to me, particularly the latter. They're unique in that they're computed by the base class and not by derived classes (although I could add virtual methods for the unlikely case that a derived class had the knowledge to make an optimised override). My instinct was to make `sortedChildNamesPlug()`  private, but that lead to some trouble because certain nodes need to make pass-through connections for it, and the ContextSanitiser needed to know about it. I could try to use `plugInputChanged()` to make the pass-throughs automatic, but that felt icky too.
- The MergeScenes implementation in this PR can clearly go in a minor release, but I'm not sure if the ScenePlug changes would be better off waiting for a major release.

Thoughts? Given the small overhead and the significant side benefits for nodes like CopyPrimitiveVariables, I'm feeling like it might be nice to follow through on the `exists()` stuff. Then again, I really need to get back to the todo list I had before I put MergeScenes at the top of it...

danieldresser-ie commented:
I do definitely find this approach much clearer, and it would be awesome to have this available for other nodes.

The 6% overhead is unfortunate, but I think it might be worth it.

Actually, if this is a private implementation detail anyway, and we just want it to answer `exists()` as fast as possible, could we cut down on the amount of plug evaluation traffic by just using one plug, sortedChildNames, which is defined to be valid to query at any location, and returns a pointer to a global empty StringVectorData at any location which doesn't exist?  That feels like it would be quite easy to define recursively the way you've done here, and cutting down on the number of plug evaluations might cut down that 6% a little bit.

> My instinct was to make sortedChildNamesPlug() private, but that lead to some trouble because certain nodes need to make pass-through connections for it

Hmm, conceptually, is this really something we want to pass through the graph?  I was kinda picturing it as something that is rarely needed, and can be computed where it's needed only when it is needed.  Philosophically, it would be nice if it was just a function on a ScenePlug, rather than being attached to a node, but we need the node so that we can use its compute machinery for the cache plugs ( or plug ).  But if we're thinking about it as a "compute on the fly when needed" thing, do we need nodes to set up pass-throughs?  I guess the question is how often do we have two nodes in a chain, where both nodes need to check exists ( pretty rare currently, but I guess it could get more common once it's faster ), and there is no node in between that touches childNames.  Maybe it would be reasonable to say that this is rare, and that no nodes should set up pass-throughs for the cache plugs, they just always get computed from the local childNames, and then it could be private?

Andrew seems pretty enthusiastic about getting MergeScenes in production hands ASAP.  Does that mean that we should merge this PR as is ( after a bit of squashing ), before doing the fast `exists()` ( I'm tempted to say we ought to be able to sneak fast `exists()`, but that's more because I want it than because it actually seems like a good idea)?  That raises the question of whether the fast `exists()` should be tackled immediately, or left for later.  I'd like to see it get done while you're in the middle of thinking about it, but I guess it's not technically urgent.

danieldresser-ie commented:
Oh, and one last comment about performance, which probably isn't at all significant:  you had the comment about finding the index of the first bit set that there should be a bit-twiddle for that.  I don't think there's anything particularly nice for doing that without an intrinsic for __builtin_ctz.

But where you could save some cycles by bit-twiddling is that you unnecessarily use `bitset::count` quite a few times.  This is a fairly expensive operation ( especially if you don't use the dedication instruction, which it sounds like you won't if you don't restrict -march to something recent ), and you're only using it for checking if all bits are 0 ( trivial ), or checking if only 1 bit is set ( which can be done with a `v && !(v & (v - 1))`  https://graphics.stanford.edu/~seander/bithacks.html#DetermineIfPowerOf2 ).

Not that this is likely to be significant, but since you mentioned bit twiddling ...

johnhaddon commented:
> Actually, if this is a private implementation detail anyway, and we just want it to answer exists() as fast as possible, could we cut down on the amount of plug evaluation traffic by just using one plug, sortedChildNames, which is defined to be valid to query at any location, and returns a pointer to a global empty StringVectorData at any location which doesn't exist?

Isn't that basically what you objected to in MappingData? That it caused childNames to be computed one level before it was actually needed by the outside world?

> That feels like it would be quite easy to define recursively the way you've done here, and cutting down on the number of plug evaluations might cut down that 6% a little bit.

If we go with this `ScenePlug::exists()` stuff then I'm much less fussed about the 6% than I am about finding the right way of presenting it in the API. I didn't put any effort into fine tuning the new MergeScenes yet so the 6% might be right there. A small constant factor is fairly irrelevant at this point anyway - the main thing is that the two approaches scale equivalently, whereas the old `SceneAlgo::exists()` mechanism scales very poorly.

> Philosophically, it would be nice if it was just a function on a ScenePlug

Definitely. I don't think `existsPlug()->hash()` is useful to anyone, and `sortedChildNames()` is just weird. They're not first-class things in the same way the other scene properties are.

> but we need the node so that we can use its compute machinery for the cache plugs ( or plug ). But if we're thinking about it as a "compute on the fly when needed" thing, do we need nodes to set up pass-throughs?

I think we should. We're optimising, and pass-throughs are a very worthwhile optimisation.

> I guess the question is how often do we have two nodes in a chain, where both nodes need to check exists ( pretty rare currently, but I guess it could get more common once it's faster ), and there is no node in between that touches childNames.

CopyAttributes followed by CopyPrimitiveVariables is a pretty obvious combo. But the nodes don't have to be in a chain either, they can be in separate branches from the same root. And a relatively low proportion of nodes touch childNames, so I don't think this scenario is as rare as you hope.

Rather than ignore the obvious optimisation, I'm becoming more inclined to make the new plugs private as we want, declare friendship with SceneNode (to be the co-conspirator that does the computes it needs that anyway), and use `plugInputChanged()` to set up the pass-throughs automatically based on `childNamesPlug()->getInput()`. I _think_ that only leaves us with `ContextSanitiser` and I wouldn't feel too bad just hardcoding a plug name in there. I'd see those as very worthwhile tradeoffs vs exposing additional plugs for the entire world to see...

> Andrew seems pretty enthusiastic about getting MergeScenes in production hands ASAP. Does that mean that we should merge this PR as is ( after a bit of squashing ), before doing the fast exists() ( I'm tempted to say we ought to be able to sneak fast exists(), but that's more because I want it than because it actually seems like a good idea)? That raises the question of whether the fast exists() should be tackled immediately, or left for later. I'd like to see it get done while you're in the middle of thinking about it, but I guess it's not technically urgent.

If you're happy with the mechanism I outlined above, then I'd be on board with pushing though `exists()` now. I've had it in mind for a while, but what has been holding me back has been unhappiness about how to expose it. If we can expose only `ScenePlug::exists( path )` I'll be pretty happy. That leaves the question of major or minor version. It doesn't break ABI, and we happily add other plugs in minor releases. I'm just being cautious because it's a very fundamental type, but maybe it's reasonable to say that 0.56 is so fresh that folks are only in a testing phase with it anyway?

danieldresser-ie commented:
> Isn't that basically what you objected to in MappingData? That it caused childNames to be computed one level before it was actually needed by the outside world?

Well, part of my objection was that I had totally forgotten about how tricky computing existence was, and the other part was that there was a lot of stuff being stored in MappingData.  If this is nicely hidden behind ScenePlug::exists(), and it doesn't require any extra data to be stored, then the cost of sticking a ParentScope inside ScenePlug::exists() seems very worth it for halving the number of plug evaluations.

> I think we should. We're optimising, and pass-throughs are a very worthwhile optimisation.
> Rather than ignore the obvious optimisation, I'm becoming more inclined to make the new plugs private as we want, declare friendship with SceneNode (to be the co-conspirator that does the computes it needs that anyway), and use plugInputChanged() to set up the pass-throughs automatically based on childNamesPlug()->getInput().

Sounds like a fair compromise.

> If you're happy with the mechanism I outlined above, then I'd be on board with pushing though exists() now. 

Sounds good.

> I'm just being cautious because it's a very fundamental type, but maybe it's reasonable to say that 0.56 is so fresh that folks are only in a testing phase with it anyway?

Yeah, if we can get a minor release out quickly, before people settle in to 0.56 too much, this seems quite reasonable, and it would keep Andrew happy.

johnhaddon commented:
OK, I've updated with a tidied up version based on the experimentation from yesterday. It is much much simpler this way, and also seems to get to within about 3% of the performance from the more complex version. I'm feeling much happier with it, so thanks for pushing me to try it.

That's the good news. The bad news is that having slept on it, I'm not happy pushing the `ScenePlug::exists()` stuff into 56. The implementation actually felt like it went pretty well, but it's in an absolutely fundamental component, and editing the graph from signals (`plugInputChanged()` in this case) always makes me nervous. So the deal is you get MergeScenes in 56 _now_ so you can start integrating it in the pipeline early, and then in 57 it magically gets 2x faster at the same time as CopyPrimitiveVariables. I'll open a PR for that on master now.

> the cost of sticking a ParentScope inside ScenePlug::exists() seems very worth it for halving the number of plug evaluations.

I didn't play with this specific idea for `ScenePlug::exists()`, but did end up doing basically the same experiment in `MergeScenes::computeActiveInputs()`. Yesterday I wasn't using the parent's active inputs to mask the `exists()` queries made for the children, and today I am. The evidence from that suggests that it's not really worth making a new context to save just one evaluation. With only two inputs (where we save one `exists()` call for most locations), it was a wash. I had to get up to four inputs (where we're saving three `exists()` calls) before it was a clear win. The balance point will obviously change when we optimise Context creation some more, but that's seems to be about where we're at now...

danieldresser-ie commented:
> The evidence from that suggests that it's not really worth making a new context to save just one evaluation.

Hmm, I'm not sure whether I'm not understanding you, or I'm not communicating my idea.

Currently, in computeExists, we must construct a parent context, and then read both existsPlug()->getValue() and sortedChildNamesPlug().  If we removed existsPlug() and made sortedChildNamesPlug() return empty when the parent doesn't exist, we would save a plug evaluation.

Oh, I guess you're considering the case where everything is already cached, and then reading an already cached value of existsPlug() is faster.  Hmm, that does make sense - I guess my concern is maybe more about the load on the value cache and hash cache from having twice as many intermediate plugs.  This thing does feel like it could put a fair bit of extra strain on caches in a scene with a deep hierarchy.  To be honest though, I probably just don't like storing two pieces of data where one will do.

johnhaddon commented:
> Hmm, I'm not sure whether I'm not understanding you, or I'm not communicating my idea.

No, you were making sense. I'd just failed to realise that what I was measuring yesterday wasn't analogous to what you were proposing. In my case I had a choice between a brute-force evaluation of several plugs, or the evaluation of a subset of them obtained via a parent context. So I was measuring the point at which the overhead of the new context justified itself.

> Currently, in computeExists, we must construct a parent context, and then read both existsPlug()->getValue() and sortedChildNamesPlug(). If we removed existsPlug() and made sortedChildNamesPlug() return empty when the parent doesn't exist, we would save a plug evaluation.

Right, so in your case the parent context is unavoidable, so it is choice a straight choice between one evaluation or two.

> Oh, I guess you're considering the case where everything is already cached, and then reading an already cached value of existsPlug() is faster.

You give me too much credit, but that is an excellent point. It does seem that interactive updates are often dominated by context construction time, so avoiding yet another temporary context is probably a good thing. Let's continue this discussion over on the relevant PR.

I've pushed a fix for the Mac build errors now, so if this LGTY then it'd be great to get it merged.

johnhaddon commented:
Pushed one more commit, that adds an obvious optimisation to `computeActiveInputs()`. This gets us to the point where we're faster than the original PR on that benchmark, even when using `SceneAlgo::exists()` for our existence queries. Note that that benchmark was for two sparsely overlapping scenes : for densely overlapping scenes the optimisation isn't effective, and `ScenePlug::exists()` is still a massive win.

danieldresser-ie commented:
I didn't end up getting much more time to look at this, but I was already OK with it, and all the changes look like definite improvements, so I'm going to say LGTM.

carstenkolve commented:
so it took a little while, but I just got to play with it a bit - very nice! one suggestion, it would be nice to  be able to provide a per input root from which onwards to perform the merging; its mostly for convenience of use; I encounter a lot of situations like

/a/BLA/x
and
/b/BLA/x/y/z

it'd be nice to be able to define
/a/BLA and  /b/BLA as respective roots
in order to receive
/a/BLA/x/y/z

its most likely possible to work around this, but it incurs a  graph clutter penalty
